cmd/util/ledger/migrations/cadence_value_validation.go

```diff
@@ -0,0 +1,496 @@
+package migrations
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/onflow/atree"
+	"github.com/onflow/cadence"
+	"github.com/onflow/cadence/runtime"
+	"github.com/onflow/cadence/runtime/common"
+	"github.com/onflow/cadence/runtime/interpreter"
+	"go.opentelemetry.io/otel/attribute"
+
+	"github.com/onflow/flow-go/cmd/util/ledger/util"
+	"github.com/onflow/flow-go/ledger"
+)
+
+var nopMemoryGauge = util.NopMemoryGauge{}
+
+// TODO: optimize memory by reusing payloads snapshot created for migration
+func validateCadenceValues(
+	address common.Address,
+	oldPayloads []*ledger.Payload,
+	newPayloads []*ledger.Payload,
+) error {
+	// Create all the runtime components we need for comparing Cadence values.
+	oldRuntime, err := newReadonlyStorageRuntime(oldPayloads)
+	if err != nil {
+		return fmt.Errorf("failed to create validator runtime with old payloads: %w", err)
+	}
+
+	newRuntime, err := newReadonlyStorageRuntime(newPayloads)
+	if err != nil {
+		return fmt.Errorf("failed to create validator runtime with new payloads: %w", err)
+	}
+
+	// Iterate through all domains and compare cadence values.
+	for _, domain := range domains {
+		err := validateStorageDomain(address, oldRuntime, newRuntime, domain)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func validateStorageDomain(
+	address common.Address,
+	oldRuntime *readonlyStorageRuntime,
+	newRuntime *readonlyStorageRuntime,
+	domain string,
+) error {
+
+	oldStorageMap := oldRuntime.Storage.GetStorageMap(address, domain, false)
+
+	newStorageMap := newRuntime.Storage.GetStorageMap(address, domain, false)
+
+	if oldStorageMap == nil && newStorageMap == nil {
+		// No storage for this domain.
+		return nil
+	}
+
+	if oldStorageMap == nil && newStorageMap != nil {
+		return fmt.Errorf("old storage map is nil, new storage map isn't nil")
+	}
+
+	if oldStorageMap != nil && newStorageMap == nil {
+		return fmt.Errorf("old storage map isn't nil, new storage map is nil")
+	}
+
+	if oldStorageMap.Count() != newStorageMap.Count() {
+		return fmt.Errorf("old storage map count %d, new storage map count %d", oldStorageMap.Count(), newStorageMap.Count())
+	}
+
+	oldIterator := oldStorageMap.Iterator(nopMemoryGauge)
+	for {
+		key, oldValue := oldIterator.Next()
+		if key == nil {
+			break
+		}
+
+		stringKey, ok := key.(interpreter.StringAtreeValue)
+		if !ok {
+			return fmt.Errorf("invalid key type %T, expected interpreter.StringAtreeValue", key)
+		}
+
+		newValue := newStorageMap.ReadValue(nopMemoryGauge, interpreter.StringStorageMapKey(stringKey))
+
+		if !cadenceValueEqual(oldRuntime.Interpreter, oldValue, newRuntime.Interpreter, newValue) {
+			return fmt.Errorf("failed to validate domain %s, key %s: old value %v (%T), new value %v (%T)", domain, key, oldValue, oldValue, newValue, newValue)
```

Comment:
What will happen here if the value is a 100GB cadence value... 

Would it be better to only log the deepest most difference instead of the outermost?

i.e. if difference is in
- composite A:
   - array B:
      -  element 2:
          -  Composite C:
              -  field D
                
we would log something like: "A.B[2].C.D (which is 'asdf') is different from A'.B'[2].C'.D' (which is 'asdfg')"
instead of entire A and entire A'

But then again, this is only relevant if we actually find a difference, so it might be premature to think about this.

Reply:
Good point!  Done in 706b5a5bb5325fe42f43f67dd2c059d0e7dfb132:

- default: modified error message to include trace and only include values at the deepest level.
- non-default flag:  `--log-verbose-validation-error` enables logging of entire Cadence values on validation error during atree migration.

For example, this is a test snippet for non-verbose error msg and verbose error log:

```Go
wantErrorMsg := "failed to validate value for address 0000000000000001, domain storage, key 0: failed to validate ([AnyStruct][0]).([UInt64][1]): values differ: 1 (interpreter.UInt64Value) != 2 (interpreter.UInt64Value)"

wantVerboseMsg := "{\"level\":\"info\",\"address\":\"0000000000000001\",\"domain\":\"storage\",\"key\":\"0\",\"trace\":\"failed to validate ([AnyStruct][0]).([UInt64][1]): values differ: 1 (interpreter.UInt64Value) != 2 (interpreter.UInt64Value)\",\"old value\":\"[[0, 1]]\",\"new value\":\"[[0, 2]]\",\"message\":\"failed to validate value\"}\n"
```